File: crawlers/db_functions.js
const {firestore, rl} = require('./data');
exports.createCollection = createCollection;
exports.getAllCollectionsAndCount = getAllCollectionsAndCount;
exports.getCollectionSize = getCollectionSize;

async function deleteCollection(collectionName, batchSize) {
    const exists = await collectionExists(collectionName);

    if (exists) {
        const collectionRef = firestore.collection(collectionName);
        const query = collectionRef.limit(batchSize);

        return new Promise((resolve, reject) => {
            deleteDocuments(query, batchSize, resolve).catch(reject);
        });
    } else {
        console.log(`Invalid collection name '${collectionName}'`);
    }
}

async function deleteDocuments(query, batchSize, resolve) {
    const dbRecords = await query.get();

    if (dbRecords.size === 0) {
        resolve();
        return;
    }

    const batch = firestore.batch();
    dbRecords.docs.forEach((doc) => {
        batch.delete(doc.ref);
    });

    await batch.commit();

    process.nextTick(() => {
        deleteDocuments(query, batchSize, resolve);
    });
}

async function deleteAllCollections() {
    const collections = await firestore.listCollections();

    for (const collection of collections) {
        console.log(`Deleting collection: ${collection.id}`);

        const documents = await collection.listDocuments();
        const deletePromises = documents.map(doc => doc.delete());
        await Promise.all(deletePromises);

        console.log(`Deleted collection: ${collection.id}`);
    }

    console.log('All collections have been deleted.');
}

async function getAllDocuments(collectionName) {
    const exists = await collectionExists(collectionName);
    if (!exists) {
        console.log(`Invalid collection name '${collectionName}'`);
        return;
    }
    const collectionRef = firestore.collection(collectionName);
    const dbRecords = await collectionRef.get();

    if (dbRecords.empty) {
        console.log('No documents in the collection.');
        return;
    }

    dbRecords.forEach(doc => {
        console.log(doc.id, '=>', doc.data());
    });
}


async function getCollectionSize(collectionName) {
    const exists = await collectionExists(collectionName)
    if (exists) {
        const collectionRef = firestore.collection(collectionName);
        const dbRecords = await collectionRef.get();
        return dbRecords.size;
    } else {
        console.log(`Invalid collection name:'${collectionName}'`);
    }

}

async function listAllCollections() {
    const collections = await firestore.listCollections();
    collections.forEach(collection => {
        console.log('Collection:', collection.id);
    });
}

async function collectionExists(collectionID) {
    const collectionRef = firestore.collection(collectionID);
    const dbRecords = await collectionRef.limit(1).get();
    return !dbRecords.empty;
}

async function getAllCollectionsAndCount() {
    try {
        const collections = await firestore.listCollections();
        const numberOfCollections = collections.length;
        console.log(`Total number of collections: ${numberOfCollections}`);

        return numberOfCollections;
    } catch (error) {
        console.error(error);
    }
}


async function getCollectionID() {
    const collectionCount = await getAllCollectionsAndCount();
    if (collectionCount !== 0) {
        const collections = await firestore.listCollections();
        collections.forEach((collection, index) => {
            console.log(`${collection.id}`);
        });

        const collectionName = await new Promise((resolve) => {
            rl.question('Choose collection: ', (answer) => {
                resolve(answer);
            });
        });

        if (collections.some(collection => collection.id === collectionName)) {
            return collectionName;
        }

    }
    return null;
}

async function createCollection() {
    let collectionName = await getCollectionID();
    if (collectionName === null) {
        collectionName = await new Promise((resolve) => {
            rl.question("Insert new collection name:", (answer) => {
                rl.close();
                resolve(answer);
            });
        });

        console.log(`Collection reference prepared for '${collectionName}'`);
    } else {
        console.log(`Using existing collection '${collectionName}'`);
    }

    return collectionName;
}


deleteAllCollections();

File: crawlers/db_functions.py
import firebase_admin
from firebase_admin import credentials, firestore
from datetime import datetime

cred = credentials.Certificate('/home/jakubg/data-db-406620-2fe1ab987b4d.json')
app = firebase_admin.initialize_app(cred)
db = firestore.client()


def check_if_collection_exist(collection):
    collection_ref = db.collection(collection)
    if collection_ref.get():
        return 1
    else:
        print(f"collection:{collection} doesn't exist")
        return 0


def insert(offer_data, collection):
    collection_ref = db.collection(collection)
    data = {
        'title': offer_data.title,
        'price': offer_data.price,
        'link': offer_data.link
    }
    collection_ref.add(data)


def insert_offers(offers, collection_name):
    collection_ref = db.collection(collection_name)
    scraping_date = datetime.now().strftime("%Y-%m-%d")

    detailed_product_data = [{'title': offer.title, 'price': offer.price, 'link': offer.link, 'provider': offer.provider} for offer in offers]

    complete_product_data = {
        'detailedProducts': detailed_product_data,
        'mainProduct': collection_name,
        'date': scraping_date
    }

    collection_ref.add(complete_product_data)





def clear_collection(collection):
    if check_if_collection_exist(collection):
        collection_ref = db.collection(collection)
        document = collection_ref.stream()
        for data in document:
            data.reference.delete()


def print_collection(collection):
    if check_if_collection_exist(collection):
        docs = db.collection("cities").stream()
        for doc in docs:
            print(f"{doc.id} => {doc.to_dict()}")


def delete_collection(collection, batch_size):
    docs = collection.list_documents(page_size=batch_size)
    deleted = 0

    for doc in docs:
        print(f"Deleting doc {doc.id} => {doc.get().to_dict()}")
        doc.delete()
        deleted = deleted + 1

    if deleted >= batch_size:
        return delete_collection(collection, batch_size)


def delete_all_collections():
    collections_list = db.collections()
    for collection_name in collections_list:
        delete_collection(collection_name, 1)


def list_All_Collections():
    collections_list = db.collections()
    for collection_name in collections_list:
        print(collection_name)

    print('All collections have been deleted.')
